xdao/gcdmaster.man
.TH GCDMASTER 1 "Oct 10, 1999"
.SH NAME
gcdmaster \- Graphical front end to cdrdao for composing audio CDs
.SH SYNOPSIS
.B gcdmaster
.RI [ toc-file ]
.SH DESCRIPTION
.I gcdmaster
allows the creation of toc-files for cdrdao and can control the
recording process. Its main application is the composition of audio
CDs from one or more audio files. It supports PQ-channel editing,
entry of meta data like ISRC codes/CD-TEXT and non destructive cut of
the audio data.

If a
.I toc-file
is specified it will be read and the referenced audio data will be displayed.
It is also possible to specify a ".cue" file.

The GUI periodically polls all configured CD-ROM and CD-recorder
devices (see section DEVICE CONFIGURE DIALOG) to retrieve their status
(ready, busy, no disk). This is done by sending a TEST UNIT READY
command to the devices. Following problems may arise:
.TP
o
Some devices (e.g. the Philips CDD2600) block the SCSI bus when a TEST
UNIT READY is issued while it logs in a new medium. This will cause a
buffer under run for all currently recording devices that are
connected to the same bus.
.TP
o
The GUI cannot detect if a device is used by another program. It will
continue to poll the device which may disturb the operation of the
other program.
.PP
The GUI supports recording of the same or different projects on
multiple devices in parallel. However, there are some caveats and your
system must meat some prerequisites:
.TP
o
Under Linux a kernel version >= 2.2.6 should be used. The generic SCSI
device of older kernels does not support parallel access to multiple
devices. 
.TP
o
The bandwidths of the disks that contain the source data and the involved
busses must be big enough to serve all recorder devices.
.TP
o
Some recorder devices may block the SCSI bus when the disk is ejected
by the software after the burning process (e.g. the Plextor PX-R412). 
For this reason a warning message will be displayed if the "Eject"
button is checked in the "Record" dialog. Manually ejecting a disk
seems not to be a problem.
.TP
o
ATAPI devices supported by the 'generic-mmc' driver will block the IDE
bus while writing the lead-in and lead-out. Thus parallel writing with
such devices connected to the same IDE channel will not work. There is
a way to avoid this blocking with some ATAPI devices but it is not
implemented, yet.

.SH CONCEPTS
The main idea of this GUI is to maintain a continuous stream of audio
data that can be composed of several audio files. It is possible to
use only portions of audio files in the audio stream which is the
basic idea for the non destructive cut capability.

For example, if you cut out some samples in the middle of an audio
file the result will be a portion that starts at the beginning of the
audio file and ends at the beginning of the cut region and a second
portion that starts at the end of the cut region and reaches until the
end of the audio file. Of course, all of this is hidden by the GUI and
you will just see the result.

Track and index marks are placed relatively to the continuous audio
stream. It is not necessary to have a separate audio file for each
track. Track/index marks can be set, moved and deleted without
influencing the audio data stream.

.SS Sample Marker
A sample marker serves as an insertion position for audio data. It is
equivalent to the cursor of a text editor. All functions that insert
audio data require a defined marker. At most one sample marker may be
active at any time.

.SS Sample Selection
A sample selection specifies a continuous range of audio samples. At
most one sample selection may be active at any time.

.SS Track Mark
A track mark specifies a point in the audio stream where the track
number or the index number changes. Track marks are written as a pair
of two numbers like 5.1 for track 5, index 1. Track numbers may be in
the range 1..99, the valid range for index numbers is 0..99. Index 0
represents the pre-gap of a track where the track relative time counts
backwards down to zero. Index 1 marks the real start of the
track. This position is stored in the central toc of the CD and is
used by CD players to directly jump to a track. All index numbers > 2
may be used to subdivide a track but have no further effect. Some CD
players may jump to such index marks.

Track marks may be selected to show data about a track or to perform
operations on the complete track. At most one track mark may be
selected at any time.

.SS Times
The time is usually displayed as m:s:f.x where m represents minutes, s
represents seconds (0..59), f represents frames (0..74, 1/75 second)
and x stands for samples (0..587, 1/44100 second).

.SH MAIN WINDOW

The main window consists of a menu bar, a sample display, a
marker/selection line, a button line and a status line.

The title of the main window shows the current toc-file name. A "(*)"
behind the name indicates that changes have not been saved.

.SS Menu Bar
The functions that are accessible via the menu bar are described in
section MENU FUNCTIONS.

.SS Sample Display
The sample display provides a visual representation of the audio
stream and the placed track/index markers. 

The samples are displayed as two separate graphs for the left and
right channel where the x-axis represents the time and the y-axis
denotes the amplitude (linearly scaled). If a pixel represents more
than 1 sample the maximum and minimum amplitude of all samples that
fall within this pixel is displayed. It is possible to zoom in and out
and change the displayed portion with the scroll bar.

The track/index markers are shown on top of the two graphs. Track
marks with index 1 are represented by filled track symbol. All other
track marks use a hollow track symbol. The track/index number pair is
displayed on the right side of a track symbol.

Track marks can be selected or moved by clicking or dragging with the
mouse pointer located over a track symbol. The track/index number
pairs are not sensitive.

.SS Marker/Selection Line

This line shows the actual cursor and active marker position and the
active sample selection.

The cursor field is read-only and shows the time value of the actual mouse
pointer position within the audio stream if it is located inside the
sample display. During playback the cursor field shows the time value
of currently audible sound.

The marker field shows the time value of the active marker. A time
value may be entered and hitting the return key sets the new marker
position if the time value is valid.

The sample selection fields show the active selected sample
range. Time values may be entered and hitting the return key in one of
these fields sets the new sample selection if the time values are valid.

.SS Button Line

The zoom/select radio buttons specify the behavior when dragging with
the mouse in the sample display. If "zoom" is active the sample
display will zoom to selected sample range. If "select" is active the
active sample selection will be set to the selected sample range.

The play button will playback the currently active sample selection
via the sound card. If no sample selection is active the currently
displayed sample range will be used. While playing the current project
is set to read-only state and all operations that would modify the
project are disabled. 

.SS Status Line

The status line shows more information about the last executed action
or an error message if an action could not be executed.

.SH MENU FUNCTIONS
.SS File->New
Starts a new project. If the current work is not saved a confirmation
box will be opened.

.SS File->Open
Opens a file selector box to select a toc-file or a .cue file that
should be read. If the current work is not saved a confirmation box
will pop up.

.SS File->Save
Saves current work as a toc-file with the current file name. 

.SS File->Save As
Opens a file sector box to select a toc-file to which the current work
is saved.

.SS File->Quit
Quits gcdmaster. If the current work is not saved a confirmation box
will pop up.

.SS View->Zoom To Selection
Zooms the sample display to the currently active sample selection. If
no sample selection is active nothing will happen.

.SS View->Zoom out
Doubles the range of displayed samples.

.SS View->Fullview
Shows all samples of the audio stream in the sample display.

.SS Edit->Cut
Removes the active selected sample range from the audio stream. The
audio file on the disk will not be touched by this operation. The
sample marker will be set so that a following "Paste" operation will
revert the effect of this operation.

If no sample selection is active nothing will happen.

.SS Edit->Paste
Pastes the samples that were previously removed with "Cut" at the
current sample marker position. The sample selection is
set to the pasted samples so that a following "Cut" operation will
revert the effect of this operation.

If no sample marker is set nothing will happen.

.SS Edit->Add Track Mark
Adds a track mark with index 1 at the active sample marker
position. The position will be rounded to the next frame boundary. If the
previous or the new track would be shorter than 4 seconds no track mask is
created.

If no sample marker is active nothing will happen.

.SS Edit->Add Index Mark
Adds a track mark with index > 1 at the active sample marker
position. The position will be rounded to the next frame
boundary. Index marks cannot be created in the pre-gap area of a
track.

If no sample marker is active nothing will happen.

.SS Edit->Add Pre-Gap
Adds a track mark with index 0 (pre-gap) at the active sample marker
position. Index 0 can only be added right before a track mark with
index 1. If the length of the previous track would be shorter than 4
seconds after inserting the pre-gap this operation is not performed.

If no sample marker is active nothing will happen.

.SS Edit->Remove Track Mark
Removes selected track mark with any index. If a track mark with index
1 is removed all track marks of this track will be removed. 

If no track marker is selected nothing will happen.

.SS Tools->Disk Info
Opens the non modal "Disk Info" dialog box. See section DISK INFO
DIALOG for more details.

.SS Tools->Track Info
Opens the non modal "Track Info" dialog box. See section TRACK INFO
DIALOG for more details.

.SS Tools->Append Track
Opens a non modal file selector dialog box to select an audio file
that will be completely appended to the audio stream. A track mark
with index 1 is added at the beginning of the appended audio data.

The non modal file selector box is shared for "Tools->Append Track",
"Tools->Append File" and "Tools->Insert File".

.SS Tools->Append File
Opens a non modal file selector dialog box to select an audio file
that will be completely appended to the audio stream.

The non modal file selector box is shared for "Tools->Append Track",
"Tools->Append File" and "Tools->Insert File".

.SS Tools->Insert File
Opens a non modal file selector dialog box to select an audio file
that will be completely inserted in the audio stream at the active
marker position. If no sample marker is active nothing will happen.

The non modal file selector box is shared for "Tools->Append Track",
"Tools->Append File" and "Tools->Insert File".

.SS Tools->Append Silence
Opens a non modal dialog box that can be used to append silence to the
audio stream.

The non model dialog box is shared for "Tools->Append Silence" and
"Tools->Insert Silence".

.SS Tools->Insert Silence
Opens a non modal dialog box that can be used to insert silence in the
audio stream at the active marker position. If no sample marker is
active nothing will happen.

The non model dialog box is shared for "Tools->Append Silence" and
"Tools->Insert Silence".

.SS Settings->Devices
Opens the non modal "Configure Devices" dialog box. See section DEVICE
CONFIGURE DIALOG for more details.

.SS Actions->Record
Opens the non modal "Record" dialog box. See section RECORD DIALOG for
more details.

.SH DISK INFO DIALOG (Tools->Disk Info)
This non modal dialog shows summary information about the whole
project and allows editing of meta data that is valid for the complete
disk.

The "Apply" button must be used to make changes permanent. If the
summary information changes before the "Apply" button was hit,
e.g. by adding a new track, all changes will be lost. 

The "Cancel" button withdraws all changes and closes the dialog box.

.SS Summary Frame
Shows the total number of tracks, i.e. number of track marks with
index 1, and the total length of the audio stream.

.SS Sub-Channel Frame
Allows entry and editing of data that is written to the sub-channels
of a disk. The scope of this data covers the whole disk.

The "Toc Type" option menu can be used to select the type of the table
of contents that will be written to the disk. Currently, for projects
that are solely created with this GUI only the type "CD-DA" or
"CD-ROM-XA" makes sense. If you read a toc-file that contains data
track specifications another type may be displayed.

The "UPC/EAN" edit field sets the catalog number of the disk. Exactly
13 digits must be entered to set the catalog number. To clear the
catalog number the edit field must be completely cleared.

.SS CD-TEXT Frame

CD-TEXT data can be specified for up to 8 different languages. Each language
is reference by a language number 0..7. The first language should have
language number 0. Language numbers should be used continuously.

The actual language for a language number is specified with the
"Language" option menu: The option "Unknown" is shown for an unknown
language code (e.g. if the toc-file was edited manually). It is not
possible to select this option. Option "Undefined" means that no
language was assigned for this language number. It is possible to
enter CD-TEXT data and the entered data will be saved to the toc-file
but it will not be recorded to a CD-R/CD-RW. The remaining options
select the desired language. Each language can be assigned only to one
language number.

The remaining CD-TEXT fields contain ASCII data. It is not necessary
to use them all but at least "Title" and "Performer" should be filled.

.SH TRACK INFO DIALOG (Tools->Track Info)
This non modal dialog shows information about a selected track and
allows editing of track specific meta data. If no track mark is
selected all fields will be cleared and set insensitive.

The "Apply" button must be used to make changes permanent. If the
track information changes, e.g. by adding a index mark, or another
track mark is selected before the "Apply" button was hit all changes
will be lost.

The "Cancel" button withdraws all changes and closes the dialog box.

.SS Summary Frame
Shows the length of the pre-gap (0 if no pre-gap is defined), the
absolute start and end time, the track length excluding the pre-gap
length and the number of index marks > 1.

.SS Sub-Channel Frame
Allows editing of all data that is written to the Q sub-channels of
the track:

.TP
Copy:
Digital copy permitted when selected.

.TP
Pre Emphasis:
Audio data of track has pre emphasis when selected.

.TP
Two/Four Channel Audio: 
Select "Four Channel Audio" if the audio data encodes more than two
channels. Of course, only two real channels will be recorded on a
CD-R/CD-RW in any case.

.TP
ISRC:
Allows to edit the ISRC code of the track. The first fields
holds the country code; it will only accept capital letters. The
second field holds the owner code, capital letters and digits are
allowed. The remaining fields holds the year and a serial number and
will only accept digits. All fields must be filled to set the ISRC
code and all fields must be cleared to remove it.

.SS CD-TEXT Frame
Allows to enter and edit the CD-TEXT data of each language number for
this track. All fields are optional but at least "Title" and
"Performer" should be filled.

.SH DEVICE CONFIGURE DIALOG (Settings->Devices)
This non modal dialog shows the settings of all attached SCSI or ATAPI
devices that identify themselves as CD-ROM or WORM device.

All device settings that differ from the default settings will be
permanently stored in the
.I $HOME/.gnome/GnomeCDMaster
configuration file.

The "Apply" button must be used to make all changes permanent. The
"Reset" button reverts all changes since the last apply or the dialog
box was opened. The "Cancel" button withdraws all changes and closes
the dialog box.

.SS Devices Frame
The browser shows the SCSI address (bus,id,lun) and the vendor/model
string of all available devices. If devices are not automatically
detected it is possible to add devices manually to the browser (see
"Add Device" Frame).

The "Status" column shows the actual device status:

.TP
Ready:
Device has a disk loaded and is ready for an action.
.TP
Busy:
Device is active with an unspecified action.
.TP
Recording:
Device is recording.
.TP
No disk:
Device has no disk loaded.
.TP
Not available:
Device cannot be accessed.
.PP

ChangeLog
--------------------------------------------------------------------------
  cdrdao 1.2.3
--------------------------------------------------------------------------

 o  By default, cdrdao tries to build with a native SCSI interface. Use
    "--with-libscg" to use bundled libscg library. (Denis Leroy)

 o  Complete rewrite of native Linux SG driver for SG 3.0 using SG_IO
    ioctl. That should make it behave better with ConsoleKit based
    permissions environment. Cdrdao should now be compiled on Linux
    without cdrtools by using "--without-scglib". (Denis Leroy)

 o  Added support for new FIRST_TRACK_NO keyword to allow a CD to start
    with a track number other than 1. (Edgar Fuss)

 o  Added native interfaces for Irix, Mac OSX and NetBSD. (Edgar Fuss)

 o  Fixed bug in Toc file parser track length computation. (Denis Leroy)

 o  Man page typo fixes. (Denis Leroy)

 o  Cygwin compilation fixes. Other various compilation warning
    fixes. (Denis Leroy)

 o  Some scsilib fixes for FreeBSD, 32-bit and 64-bit SPARC. (Marius
    Strobl)

 o  Some makefile changes to avoid conflict with installed system libs.
    (Marius Strobl)

 o  Added patch to fix ARM and MIPS scsilib compile (Kedar Sovani, Debian)

 o  Added man pages for toc2cue and toc2cddb, from Debian.

 o  Move privilege drop earlier in main flow (based on Debian patch)

GCDMaster changes:

 o  Preferences now stored in GConf, exports schema file (Denis Leroy)

 o  Added simple Preferences dialog to change temporary file directory
    (Denis Leroy)

 o  Fixed GCC 4.3 compilation issues (Denis Leroy)

--------------------------------------------------------------------------
  cdrdao 1.2.2 (2006-09-20)
--------------------------------------------------------------------------

 o  Added device locking for win32 (Giuseppe Corbelli)

 o  Man page cleanup (Andrew Suffield)

GCDMaster changes:

 o  Fixed GCC 4.1 compilation issues

 o  Fixed hang when inserting WAV file


--------------------------------------------------------------------------
  cdrdao 1.2.1 (2005-10-28)
--------------------------------------------------------------------------

 o  Fixed bad bug with missing return statemements in recomputeLength
    and format conversion.

 o  Fixed broken '-' STDIN input feature.

 o  Fixed various gcc4-related compile problems.

 o  Fixed copy problem with single device when read_device is
    specified.

 o  Added autodetect of cdrtools header files in configure

 o  Minor fixes to driver table.

 o  pccts build is now better integrated with automake/autoconf.

 o  Added FreeBSD 7 support to configure.ac. Fixed libao dependency.	

GCDMaster changes:

 o  The crash when creating a new "Create Audio CD" view seemed to be
    related to gtkmm and libsigc++ and went away with newer releases of
    those packages.

toc2cue changes:

 o  toc2cue: fixed bug that was causing empty output files.

 o  Added conversion of COPY to DCP flags. (Julio Sanchez Fernandez)


--------------------------------------------------------------------------
  cdrdao 1.2.0 (2005-05-12)
--------------------------------------------------------------------------

 o  SECURITY FIX: cdrdao now gives up its root privileges after setting
    up real-time scheduling, as well as before saving settings through
    the --save option. This fixes a potential local root exploit when
    cdrdao is installed with the +s chmod flag. Using --save now also
    forces an early exit after the settings are saved.

 o  Added MP3 and Ogg Vorbis file support (through respectively the
    libmad and libvorbis libraries). cdrdao will decode the MP3 and
    Ogg files into temporary WAV files that will be deleted upon exit
    (unless the new option --keep is used). The directory used to
    store those temporary WAV files can be specified with the --tmpdir
    option (default is /tmp).

 o  Improved native CUE file support: replaced old incomplete existing
    parser with the one from Matthias Czapla's excellent
    cue2toc. Added support for cutting binary files.

 o  Added --rspeed option to manually set the device reading speed. Be
    warned not all devices support this.

 o  Packaged scsilib library upgraded from cdrtools 2.01 (previously
    was from 2.01a31).

 o  Added --no-mode2-mixed option. Don't read a MODE2 disk as
    MODE2_FORM_MIXED, only read MODE2_FORM1 or MODE2_FORM2 (Matthieu
    Castet).

 o  Added help for little-known drive-info command.

GCDMaster changes:

 o  MP3 and Ogg Vorbis support: you can drag and drop .mp3, .m3u and
    .ogg files from Nautilus into the sample display window.

 o  Switched to gtkmm24 API. Improved file browsers.

 o  CUE files support.

 o  Sound output now uses libao library.

 o  Added 'Select All' item in menu.

 o  Added 'Eject' button to progress dialog

 o  Bug fixes (sample selection weirdness, couldn't close window during
    play, problems with gcdmaster command-line argument, crashes with
    multiple project windows, drive status not reported correctly).

--------------------------------------------------------------------------
  cdrdao 1.1.9 (2004-06-07)
--------------------------------------------------------------------------

 o MMC-driver auto-detection. If no --driver option is specified,
   cdrdao will try to autodetect an MMC-compliant device and use the
   generic-mmc driver by default. (Denis Leroy)

 o Included scsilib upgraded from cdrtools 2.01a31 (Denis Leroy)

 o Added --eject option to the unlock command to eject the cd after
   unlocking. (Eric Benson)

 o New 'discid' command to display CDDB information. (Eric Benson,
   Denis Leroy).

 o Fixed scsilib compiling problems with Linux kernel 2.6 (Jim Gifford)

 o Fixed compiling problems with gcc 3.4 (Jim Gifford, Denis Leroy)

 o Added --full-burn and --capacity options to extend lead-out to entire disc,
   beyond 80 mins limit. (Vladimir Petrov)


GCDMaster changes (Denis Leroy):

 o Device configuration dialog now scans for devices on both ATA and

xdao/gcdmaster.cc
/*  cdrdao - write audio CD-Rs in disc-at-once mode
 *
 *  Copyright (C) 1998-2000  Andreas Mueller <mueller@daneb.ping.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <gtkmm.h>
#include <gnome.h>

#include "config.h"

#include "xcdrdao.h"
#include "guiUpdate.h"
#include "DeviceConfDialog.h"
#include "ProjectChooser.h"
#include "gcdmaster.h"
#include "TocEdit.h"
#include "util.h"
#include "AudioCDProject.h"
#include "DuplicateCDProject.h"
#include "BlankCDDialog.h"
#include "DumpCDProject.h"
#include "Icons.h"

// Static class members
std::list<GCDMaster *> GCDMaster::apps;

GCDMaster::GCDMaster() : Gnome::UI::App("gcdmaster", APP_NAME)
{
  set_title(APP_NAME);

  project_number = 0;
  about_ = 0;
  project_ = 0;
  chooser_ = 0;

  set_resizable();
  set_wmclass("gcdmaster", "GCDMaster");

  readFileSelector_ =
    new Gtk::FileChooserDialog(_("Please select a project"),
                               Gtk::FILE_CHOOSER_ACTION_OPEN);
  readFileSelector_->set_transient_for(*this);
  readFileSelector_->add_button(Gtk::Stock::CANCEL, Gtk::RESPONSE_CANCEL);
  readFileSelector_->add_button(Gtk::Stock::OPEN, Gtk::RESPONSE_OK);

  Gtk::FileFilter* filter_tocs = new Gtk::FileFilter;
  manage(filter_tocs);
#ifdef HAVE_MP3_SUPPORT
  filter_tocs->set_name("Content Files (*.toc, *.cue, *.m3u)");
#else
  filter_tocs->set_name("Content Files (*.toc, *.cue)");
#endif
  filter_tocs->add_pattern("*.toc");
  filter_tocs->add_pattern("*.cue");
#ifdef HAVE_MP3_SUPPORT
  filter_tocs->add_pattern("*.m3u");
#endif
  readFileSelector_->add_filter(*filter_tocs);

  Gtk::FileFilter* filter_all = new Gtk::FileFilter;
  manage(filter_all);
  filter_all->set_name("Any files");
  filter_all->add_pattern("*");
  readFileSelector_->add_filter(*filter_all);

  Icons::registerStockIcons();
  notebook_.set_show_border(false);
  set_contents(notebook_);

  createMenus();
  createStatusbar();

  apps.push_back(this);
}

void GCDMaster::createMenus()
{
  //Define the actions:
  m_refActionGroup = Gtk::ActionGroup::create("Actions");

  // File
  m_refActionGroup->add( Gtk::Action::create("FileMenu", _("_File")) );
  m_refActionGroup->add( Gtk::Action::create("New", Gtk::Stock::NEW),
                         sigc::mem_fun(*this, &GCDMaster::newChooserWindow) );

  // File->New
  m_refActionGroup->add( Gtk::Action::create("FileNewMenu", _("N_ew")) );
  m_refActionGroup->add( Gtk::Action::create("NewAudioCD", Gtk::Stock::NEW,
                         _("_Audio CD"),
                         _("New Audio CD")),
                         sigc::mem_fun(*this, &GCDMaster::newAudioCDProject2) );

  m_refActionGroup->add( Gtk::Action::create("NewDuplicateCD", Gtk::Stock::NEW,
                         _("_Duplicate CD"),
                         _("Make a copy of a CD")),
                         sigc::mem_fun(*this, &GCDMaster::newDuplicateCDProject) );

  m_refActionGroup->add( Gtk::Action::create("NewDumpCD", Gtk::Stock::NEW,
                         _("_Copy CD to disk"),
                         _("Dump CD to disk")),
                         sigc::mem_fun(*this, &GCDMaster::newDumpCDProject) );

  m_refActionGroup->add( Gtk::Action::create("Open", Gtk::Stock::OPEN),
                         sigc::mem_fun(*this, &GCDMaster::openProject) );

  m_refActionGroup->add( Gtk::Action::create("Close", Gtk::Stock::CLOSE),
                         sigc::hide_return(sigc::mem_fun(*this, &GCDMaster::closeProject)));

  m_refActionGroup->add( Gtk::Action::create("Quit", Gtk::Stock::QUIT),
                         &GCDMaster::appClose);

  // Edit
  m_refActionGroup->add( Gtk::Action::create("EditMenu", _("_Edit")) );

  // Actions menu
  m_refActionGroup->add( Gtk::Action::create("ActionsMenu", _("_Actions")) );
  m_refActionGroup->add( Gtk::Action::create("BlankCD", Gtk::Stock::CDROM,
                         _("Blank CD-RW"),
                         _("Erase a CD-RW")),
                         sigc::mem_fun(*this, &GCDMaster::blankCDRW) );

  // Settings
  m_refActionGroup->add( Gtk::Action::create("SettingsMenu", _("_Settings")) );
  m_refActionGroup->add( Gtk::Action::create("ConfigureDevices", Gtk::Stock::PREFERENCES,
                         _("Configure Devices..."),
                         _("Configure the read and recording devices")),
                         sigc::mem_fun(*this, &GCDMaster::configureDevices) );

  // Help
  m_refActionGroup->add( Gtk::Action::create("HelpMenu", _("_Help")) );
//FIXME: llanero Gtk::Stock::ABOUT ???
  m_refActionGroup->add( Gtk::Action::create("About", _("About")),
                         sigc::mem_fun(*this, &GCDMaster::aboutDialog) );

  m_refUIManager = Gtk::UIManager::create();
  m_refUIManager->insert_action_group(m_refActionGroup);
  add_accel_group(m_refUIManager->get_accel_group());

  //Layout the actions in a menubar and toolbar:
  try
  {
    Glib::ustring ui_info =
        "<ui>"
        "  <menubar name='MenuBar'>"
        "    <menu action='FileMenu'>"
        "      <menuitem action='New'/>"
        "      <menu action='FileNewMenu'>"
        "        <menuitem action='NewAudioCD'/>"
        "        <menuitem action='NewDuplicateCD'/>"
        "        <menuitem action='NewDumpCD'/>"
        "      </menu>"
        "      <menuitem action='Open'/>"
        "      <placeholder name='FileSaveHolder'/>"
        "      <separator/>"
        "      <menuitem action='Close'/>"
        "      <menuitem action='Quit'/>"
        "    </menu>"
        "    <menu action='EditMenu'/>"
        "    <menu action='ActionsMenu'>"
        "      <placeholder name='ActionsRecordHolder'/>"
        "      <menuitem action='BlankCD'/>"
        "      <separator/>"
        "    </menu>"
        "    <menu action='SettingsMenu'>"
        "      <menuitem action='ConfigureDevices'/>"
        "    </menu>"
        "    <menu action='HelpMenu'>"
        "      <menuitem action='About'/>"
        "    </menu>"
        "  </menubar>"
        "  <toolbar name='ToolBar'>"
        "    <toolitem action='New'/>"
        "    <toolitem action='Open'/>"
        "  </toolbar>"
        "</ui>";

    m_refUIManager->add_ui_from_string(ui_info);
  }
  catch(const Glib::Error& ex)
  {
    std::cerr << "building menus failed: " <<  ex.what();
  }

  Gtk::Widget* pMenuBar = m_refUIManager->get_widget("/MenuBar");
  set_menus(dynamic_cast<Gtk::MenuBar&>(*pMenuBar));
  Gtk::Widget* pToolbar = m_refUIManager->get_widget("/ToolBar");
  set_toolbar(dynamic_cast<Gtk::Toolbar&>(*pToolbar));
}

bool GCDMaster::openNewProject(const char* s)
{
  TocEdit* tocEdit;

  if (s == NULL || *s == 0 || s[strlen(s) - 1] == '/')
    return false;

  FileExtension type = fileExtension(s);
  switch (type) {

  case FE_M3U:
    newAudioCDProject("", NULL, s);
    break;

  case FE_TOC:
    tocEdit = new TocEdit(NULL, NULL);
    if (tocEdit->readToc(stripCwd(s)) == 0)
      newAudioCDProject(stripCwd(s), tocEdit);
    else
      return false;
    break;

  case FE_CUE:
    tocEdit = new TocEdit(NULL, NULL);
    if (tocEdit->readToc(stripCwd(s)) == 0)
      newAudioCDProject("", tocEdit);
    else
      return false;
    break;

  default:
    printf("Could not open \"%s\": format not supported.\n", s);
    return false;
    break;
  }

  return true;
}

void GCDMaster::openProject()
{
  readFileSelector_->present();
  int result = readFileSelector_->run();
  readFileSelector_->hide();

  if (result == Gtk::RESPONSE_OK) {
    std::string s = readFileSelector_->get_filename();
    openNewProject(s.c_str());
  }
}

bool GCDMaster::closeProject()
{
  if (chooser_)
    closeChooser();

  if (project_) {
    if (project_->closeProject()) {
      delete project_;
      project_ = NULL;
    } else {
      return false; // User clicked on cancel
    }
  }

  if (readFileSelector_)
    delete readFileSelector_;

  apps.remove(this);
  delete this;

  if (apps.size() == 0)
    Gnome::Main::quit(); // Quit if there are not remaining windows

  return true;
}

void GCDMaster::closeChooser()
{
  delete chooser_;
  chooser_ = NULL;
}

bool GCDMaster::on_delete_event(GdkEventAny* e)
{
  closeProject();
  return true;
}

// Application Close. Called when the user selects 'Quit' from the
// menu. Try to close all project windows and quit.

void GCDMaster::appClose()
{
  // Can't just iterate, as closeProject will remove its object from
  // the list.
  while (apps.size() > 0) {
    
    if (!(*(GCDMaster::apps.begin()))->closeProject())
      return;
  }

  return;
}

void GCDMaster::newChooserWindow()
{
  if (project_ || chooser_) {

    GCDMaster *gcdmaster = new GCDMaster;
    gcdmaster->newChooserWindow();
    gcdmaster->show();

  } else {

    chooser_ = new ProjectChooser();
    chooser_->newAudioCDProject.connect(sigc::mem_fun(*this,
                                        &GCDMaster::newAudioCDProject2));
    chooser_->newDuplicateCDProject.connect(sigc::mem_fun(*this,
                                        &GCDMaster::newDuplicateCDProject));
    chooser_->newDumpCDProject.connect(sigc::mem_fun(*this,
                                        &GCDMaster::newDumpCDProject));
    chooser_->show();
    notebook_.set_show_tabs(false);
    notebook_.append_page(*chooser_);
  }
}

void GCDMaster::newAudioCDProject(const char *name, TocEdit *tocEdit,
                                  const char* tracks)
{
  if (!project_) {

    AudioCDProject* p = new AudioCDProject(project_number++, name, tocEdit,
                                           this);
    p->add_menus(m_refUIManager);
    p->configureAppBar(statusbar_, progressbar_, progressButton_);

    project_ = p;
    project_->show();
    if (chooser_)
      closeChooser();
    notebook_.remove_page();
    notebook_.set_show_tabs(false);
    notebook_.append_page(*project_);
    if (tracks)
      p->appendTrack(tracks);

  } else {

    GCDMaster *gcdmaster = new GCDMaster;
    gcdmaster->newAudioCDProject(name, tocEdit, tracks);
    gcdmaster->show();
  }
}

void GCDMaster::newAudioCDProject2()
{
  newAudioCDProject("", NULL);
}

void GCDMaster::newDuplicateCDProject()
{
  if (!project_) {

    project_ = new DuplicateCDProject(this);
    project_->show();
    if (chooser_)
      closeChooser();
    notebook_.remove_page();
    notebook_.set_show_tabs(false);
    notebook_.append_page(*project_);
    set_title(_("Duplicate CD"));

  } else {

    GCDMaster *gcdmaster = new GCDMaster;
    gcdmaster->newDuplicateCDProject();
    gcdmaster->show();
  }
}

void GCDMaster::newDumpCDProject()
{
  if (!project_) {

    project_ = new DumpCDProject(this);
    project_->show();
    if (chooser_)
      closeChooser();
    notebook_.remove_page();
    notebook_.set_show_tabs(false);
    notebook_.append_page(*project_);
    set_title(_("Dump CD to disk"));

  } else {

    GCDMaster *gcdmaster = new GCDMaster;
    gcdmaster->newDumpCDProject();
    gcdmaster->show();
  }
}

void GCDMaster::update(unsigned long level)
{
  if (project_)
    project_->update(level);

  blankCDDialog_.update(level);
}

void GCDMaster::configureDevices()
{
  deviceConfDialog->start();
}

void GCDMaster::blankCDRW()
{
  blankCDDialog_.start(*this);
}

void GCDMaster::createStatusbar()
{
  Gtk::HBox *container = new Gtk::HBox;
  statusbar_ = new Gnome::UI::AppBar(false, true,
                                     Gnome::UI::PREFERENCES_NEVER);
  progressbar_ = new Gtk::ProgressBar;
  progressButton_ = new Gtk::Button(_("Cancel"));
  progressButton_->set_sensitive(false);

  progressbar_->set_size_request(150, -1);
  container->pack_start(*statusbar_, true, true); 
  container->pack_start(*progressbar_, false, false); 
  container->pack_start(*progressButton_, false, false); 
  set_statusbar_custom(*container, *statusbar_);
  container->set_spacing(2);
  container->set_border_width(2);
  container->show_all();
  install_menu_hints();
}

void GCDMaster::aboutDialog()
{
  if (about_) {
      // "About" dialog hasn't been closed, so just raise it
      about_->present();

  } else {

    std::vector<std::string> authors;
    authors.push_back("Andreas Mueller <mueller@daneb.ping.de>");
    authors.push_back("Manuel Clos <llanero@jazzfree.com>");
    authors.push_back("Denis Leroy <denis@poolshark.org>");
    std::vector<std::string> comments;

    about_ = new Gnome::UI::About("gcdmaster", VERSION,
                                  "(C) Andreas Mueller",
                                  authors, comments);

    about_->set_transient_for(*this);
    about_->show();
  }
}
